### PR TITLE
Feature: Deployment History - fetching YAMLs lazily using new API

### PR DESCRIPTION
**Pull request by sohel-rp**
# Description
Breakdown deployment history API to solve grpc max size error.

- Fetching YAMLs lazily instead of on page load using new API - `application/deployment-detail`.
- Updates in deployment tabs - added `values.yaml` & renamed `Applied YAML` to `Helm generated manifest`.
- Added error handling view containing CTA - 'Retry'.

Fixes # https://github.com/devtron-labs/devtron/issues/1155

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
All tests have been performed manually on the local server.

- [x] No manifest details are fetched on page load
- [x] YAMLs for both `values.yaml` & `Helm generated manifest` are being fetched on click of either tab.
- [x] No multiple API calls are being made while switching between `values.yaml` & `Helm generated manifest` tabs.
- [x] No API call is being made if YAMLs have already been fetched for a particular Deployment version.
- [x] Used network request blocking to block the request to new API & test out Error handling view.

# Checklist:

* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have performed a self-review of my own code
* [ ] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

File: src/components/common/icons/Progressing.tsx [hunk 1]
```diff
@@ -5,12 +5,13 @@ interface ProgressingProps {
...
     fullHeight?: boolean;
+    theme?: 'white' | 'default';
 }
 
-export function Progressing({ pageLoader, size }: ProgressingProps): JSX.Element {
+export function Progressing({ pageLoader, size, theme }: ProgressingProps): JSX.Element {
     const loaderSize = size ? `${size}px` : pageLoader ? '48px' : '20px';
     return (
-        <div className="loader">
+        <div className={`loader ${theme || 'default'}-background`}>
```
Comment (shivani170) [inline src/components/common/icons/Progressing.tsx:14]: Is `default-background`classname adding any value here?

Comment (sohel-rp) [reply src/components/common/icons/Progressing.tsx:14]: It's just a placeholder to stop including `undefined-background` as a class name :)

**Review by shivani170 — APPROVED** [review]

File: src/components/external-apps/ExternalAppService.ts [hunk 2]
```diff
@@ -12,6 +12,14 @@ export interface HelmAppDeploymentHistoryResponse extends ResponseType {
     result?: HelmAppDeploymentHistory
 }
 
+export interface HelmAppDeploymentDetails {
```
Comment (manish-agrawal-ai) [inline src/components/external-apps/ExternalAppService.ts:15]: There is another interface "HelmAppDeploymentDetail" in same file. So this name is confusing. 
You can use "HelmAppDeploymentManifestDetail"

File: src/components/v2/common/message.ui.tsx [hunk 3]
```diff
@@ -52,20 +58,25 @@ const MessageUI: React.FC<MsgUIProps> = ({
...
             {isShowActionButton && (
                 <div
                     className="cursor"
                     onClick={onActionButtonClick}
-                    style={{ fontSize: '14px', textDecoration: 'underline', color: 'var(--B300)' }}
+                    style={{
+                        fontSize: '14px',
+                        textDecoration: 'underline',
+                        color: 'var(--B300)',
+                        ...actionButtonStyle,
```
Comment (manish-agrawal-ai) [inline src/components/v2/common/message.ui.tsx:78]: if actionButtonStyle is not provided (i.e. null), please test that part also. (i.e. regression)

File: src/components/common/icons/Progressing.tsx (hunk 1 above)
Comment (manish-agrawal-ai) [inline src/components/common/icons/Progressing.tsx:14]: There is no default-background class css.

File: src/components/v2/deployment-history/ea/EADeploymentHistory.component.tsx [hunk 4]
```diff
@@ -1,30 +1,45 @@
...
+import MessageUI from '../../common/message.ui';
 
-function ExternalAppDeploymentHistory({appId}) {
+function ExternalAppDeploymentHistory({ appId }: { appId: string }) {
     const [isLoading, setIsLoading] = useState(true);
     const [errorResponseCode, setErrorResponseCode] = useState(undefined);
     const [deploymentHistoryArr, setDeploymentHistoryArr] = useState<HelmAppDeploymentDetail[]>([]);
     const [selectedDeploymentHistoryIndex, setSelectedDeploymentHistoryIndex] = useState<number>(0);
     const [selectedDeploymentTabIndex, setSelectedDeploymentTabIndex] = useState<number>(0);
+    const [deploymentDetails, setDeploymentDetails] = useState<Record<number, HelmAppDeploymentDetails>>();
```
Comment (manish-agrawal-ai) [inline src/components/v2/deployment-history/ea/EADeploymentHistory.component.tsx:27]: versionVsDeploymentManifestDetail

File: src/components/v2/deployment-history/ea/EADeploymentHistory.component.tsx [hunk 5]
```diff
@@ -1,30 +1,45 @@
...
+    const [selectedDeploymentDetails, setSelectedDeploymentDetails] = useState<HelmAppDeploymentDetails>();
```
Comment (manish-agrawal-ai) [inline src/components/v2/deployment-history/ea/EADeploymentHistory.component.tsx:28]: selectedDeploymentManifestDetail

File: src/components/v2/deployment-history/ea/EADeploymentHistory.component.tsx [hunk 6]
```diff
@@ -34,19 +49,57 @@ function ExternalAppDeploymentHistory({appId}) {
...
-        if(selectedDeploymentHistoryIndex == index){
+    function changeDeployment(index: number) {
+        if (selectedDeploymentHistoryIndex == index) {
             return;
         }
 
         setSelectedDeploymentHistoryIndex(index);
+
+        // Resetting the deployment tab selection, loading & error states on version change.
+        setSelectedDeploymentTabIndex(0);
```
Comment (manish-agrawal-ai) [inline src/components/v2/deployment-history/ea/EADeploymentHistory.component.tsx:60]: Purpose of this line?

File: src/components/v2/common/message.ui.tsx (hunk 3 above)
Comment (sohel-rp) [reply src/components/v2/common/message.ui.tsx:78]: Yes. It's already been tested and it works perfectly fine without any errors.

File: src/components/common/icons/Progressing.tsx (hunk 1 above)
Comment (sohel-rp) [reply src/components/common/icons/Progressing.tsx:14]: It's just a placeholder for default values. By default, it picks parents' background color.

File: src/components/v2/deployment-history/ea/EADeploymentHistory.component.tsx (hunk 6 above)
Comment (sohel-rp) [reply src/components/v2/deployment-history/ea/EADeploymentHistory.component.tsx:60]: For resetting deployment tab selection after changing the version. As it was maintaining the tab state from old version selection, displaying YAML of the old version.

File: src/components/v2/deployment-history/ea/EADeploymentHistory.component.tsx (hunk 5 above)
Comment (sohel-rp) [reply src/components/v2/deployment-history/ea/EADeploymentHistory.component.tsx]: changed the impl so it's been removed.

**Comment by sohel-rp** [comment]
@manish-agrawal-ai Updated as per comments & offline discussion :)

File: src/components/external-apps/ExternalAppService.ts [hunk 7]
```diff
@@ -12,6 +12,14 @@ export interface HelmAppDeploymentHistoryResponse extends ResponseType {
     result?: HelmAppDeploymentHistory
 }
 
+export interface HelmAppDeploymentManifestDetail {
+    manifest?: string;
```
Comment (manish-agrawal-ai) [inline src/components/external-apps/ExternalAppService.ts:16]: remove manifest from HelmAppDeploymentDetail as it is not being used now

File: src/components/v2/deployment-history/ea/EADeploymentHistory.component.tsx [hunk 8]
```diff
@@ -58,32 +69,36 @@ function ExternalAppDeploymentHistory({ appId }: { appId: string }) {
...
-                setErrorLoadingYaml(true);
+    async function fetchDeploymentDetail(version: number): Promise<void> {
+        try {
+            const { result } = await getDeploymentManifestDetails(appId, version);
+            initDeploymentManifestDetails.set(version, {
+                manifest: result.manifest,
+                valuesYaml: result.valuesYaml,
+                loading: false,
+                error: false,
+                isApiCallTriggered: true,
```
Comment (manish-agrawal-ai) [inline src/components/v2/deployment-history/ea/EADeploymentHistory.component.tsx:82]: isApiCallTriggered: false

File: src/components/v2/deployment-history/ea/EADeploymentHistory.component.tsx [hunk 9]
```diff
@@ -58,32 +69,36 @@ function ExternalAppDeploymentHistory({ appId }: { appId: string }) {
...
             });
+            setDeploymentManifestDetails(new Map<number, DeploymentManifestDetail>(initDeploymentManifestDetails));
+        } catch (e) {
+            initDeploymentManifestDetails.set(version, {
+                loading: false,
+                error: true,
+                isApiCallTriggered: true,
```
Comment (manish-agrawal-ai) [inline src/components/v2/deployment-history/ea/EADeploymentHistory.component.tsx:89]: isApiCallTriggered: false

File: src/components/external-apps/ExternalAppService.ts (hunk 7 above)
Comment (sohel-rp) [reply src/components/external-apps/ExternalAppService.ts:16]: Done

File: src/components/v2/deployment-history/ea/EADeploymentHistory.component.tsx (hunk 9 above)
Comment (sohel-rp) [reply src/components/v2/deployment-history/ea/EADeploymentHistory.component.tsx]: Impl changed.

File: src/components/v2/deployment-history/ea/EADeploymentHistory.component.tsx (hunk 8 above)
Comment (sohel-rp) [reply src/components/v2/deployment-history/ea/EADeploymentHistory.component.tsx]: Impl changed

**Review by manish-agrawal-ai — APPROVED** [review]

**Review by manish-agrawal-ai — APPROVED** [review]
approved

**Pull request merged**